packages/beacon-node/src/chain/prepareNextSlot.ts

```diff
@@ -144,7 +145,29 @@ export class PrepareNextSlotScheduler {
       if (isExecutionStateType(prepareState)) {
         const proposerIndex = prepareState.epochCtx.getBeaconProposer(prepareSlot);
         const feeRecipient = this.chain.beaconProposerCache.get(proposerIndex);
+        let updatedPrepareState = prepareState;
+        let updatedHeadRoot = headRoot;
+
         if (feeRecipient) {
+          // If we are proposing next slot, we need to predict if we can proposer-boost-reorg or not
+          const {slot: proposerHeadSlot, blockRoot: proposerHeadRoot} = this.chain.predictProposerHead(clockSlot);
+
+          // If we predict we can reorg, update prepareState with proposer head block
+          if (proposerHeadRoot !== headRoot || proposerHeadSlot !== headSlot) {
+            this.logger.verbose("Weak head detected. May build on this block instead:", {
```

Comment:
this `if` clause should come from the result of `chain.predictProposerHead()` above
`regen.getBlockSlotState()` should be called once in this flow
also would be great to also log the weak head

Reply:
@twoeths The thing is we need `regen.getBlockSlotState()` to get the state in order to know if we are proposing next slot.

We only `predictProposerHead()` and determine the correct `updatedPrepareState` (by calling `regen.getBlockSlotState()` again) after we know we are proposing next slot

Reply:
Also we call `regen.getBlockSlotState()` twice only if we are re-orging, which I believe wouldn't happen too often.

Reply:
I see, this is not ideal through. could you:
- add a `TODO: figure out how to call regen.getBlockSlotState() once` => I will look into this
- make sure we always call `regen.getBlockSlotState()` once if we disable the feature flag, this is to make sure there is no side affect in the mean time
- only call `state.hashTreeRoot()` once we determine the final state to prepare (line 122 above)

for the context, running an epoch transition could be up to 2s, `hash_tree_root()` could be up to > 1.5s on mainnet

Reply:
> make sure we always call regen.getBlockSlotState() once if we disable the feature flag, this is to make sure there is no side affect in the mean time

I believe it's already doing it.
If we disable the feature flag, `predictProposerHead()` will return `headRoot` and `headSlot` (see [forkChoice.ts:L228](https://github.com/ChainSafe/lodestar/blob/2f23cd7a480265cfb3c8ffa3a744a0b5c7c99fdf/packages/fork-choice/src/forkChoice/forkChoice.ts#L228) . And thus the `if` clause will not be triggered:
```
if (proposerHeadRoot !== headRoot || proposerHeadSlot !== headSlot){
    ...
    this.chain.regen.getBlockSlotState();
    ....
}
```

> only call state.hashTreeRoot() once we determine the final state to prepare (line 122 above)

Thanks for catching that. Updated.